### ISSUE TITLE
check-requires drops requires whose imported bindings are used in module+

### ISSUE THREAD

**Issue by sorawee** [racket/racket#2888]
Consider:

```
#lang racket/base

(require racket/pretty)

(module+ main
  (pretty-write 1))
```

`raco check-requires test.rkt` results in:

```
(file "test.rkt"):
DROP racket/pretty at 0
```
which is obviously incorrect.

(Not sure if this is the same bug as #2546)

**Comment by jackfirth** [comment]
I think `check-requires` is _partially_ correct there, as in my opinion that code should be changed to this:

```racket
#lang racket/base

(module+ main
  (require racket/pretty))

;; other code here

(module+ main
  (pretty-write 1))
```

**Comment by sorawee** [comment]
You could imagine:

```
#lang racket/base

(require racket/pretty)

(module+ mod1
  (pretty-write 1))

(module+ mod2
  (pretty-write 2))

(module+ mod3
  (pretty-write 3))

...
```

It makes sense in this case to `(require racket/pretty)` once at the top.

**Comment by jackfirth** [comment]
It's a bad idea to try and "deduplicate" dependency lists in code. When there are shared deps it's more often than a not a coincidence, and the deduplication encourages the proliferation of unnecessary dependencies. It also makes tooling for dependency management much more complex.

**Comment by greghendershott** [comment]
I mostly agree with you @jackfirth as a matter of style. And if it were `raco style-linter` this could be closed. :smile:

But to the extent `raco check-requires` means "I will tell you requires you may safely remove because they're unused", this seems like a legitimate issue?

---

p.s. I only "mostly" agree because I feel like wanting to "de-dupe" dependencies is not necessarily a bad thing. In fact it seems not uncommon in Racket to define a module that "only" re-provides things from other files. See also `#lang reprovide` to DRY the DRY-ing.

And if this leads to a proliferation of unnecessary dependencies, well, a tool to help find those is exactly what this issue is concerned with?

**Comment by jackfirth** [comment]
Yes, this is definitely a legitimate bug in `raco check-requires` - the require cannot be removed, so the tool shouldn't suggest removing it.
